File: backend/sim-backend/src/main/java/com/sim/backend/product/service/ProductMaterialService.java
package com.sim.backend.product.service;

import com.sim.backend.product.dto.ProductMaterialQuantityRequest;
import com.sim.backend.product.dto.ProductMaterialRequest;
import com.sim.backend.product.dto.ProductMaterialResponse;

import java.util.List;
import java.util.UUID;

public interface ProductMaterialService {

    ProductMaterialResponse addMaterial(
            UUID productId,
            ProductMaterialRequest request
    );

    List<ProductMaterialResponse> findByProductId(UUID productId);

    ProductMaterialResponse updateBaseQuantity(
            UUID productId,
            UUID productMaterialId,
            ProductMaterialQuantityRequest request
    );

    void removeMaterial(
            UUID productId,
            UUID productMaterialId
    );
}

File: frontend/src/app/features/products/product-material.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import {
  ProductMaterial,
  ProductMaterialQuantityRequest,
  ProductMaterialRequest
} from '../../shared/models/product-material';

@Injectable({
  providedIn: 'root'
})
export class ProductMaterialService {

  private readonly http = inject(HttpClient);

  private readonly apiUrl =
    'http://localhost:8080/api/products';

  findByProductId(
    productId: string
  ): Observable<ProductMaterial[]> {

    return this.http.get<ProductMaterial[]>(
      `${this.apiUrl}/${productId}/materials`
    );
  }

  addMaterial(
    productId: string,
    request: ProductMaterialRequest
  ): Observable<ProductMaterial> {

    return this.http.post<ProductMaterial>(
      `${this.apiUrl}/${productId}/materials`,
      request
    );
  }

  updateBaseQuantity(
    productId: string,
    productMaterialId: string,
    request: ProductMaterialQuantityRequest
  ): Observable<ProductMaterial> {

    return this.http.patch<ProductMaterial>(
      `${this.apiUrl}/${productId}/materials/${productMaterialId}`,
      request
    );
  }

  removeMaterial(
    productId: string,
    productMaterialId: string
  ): Observable<void> {

    return this.http.delete<void>(
      `${this.apiUrl}/${productId}/materials/${productMaterialId}`
    );
  }
}